### ISSUE TITLE
IOException might not be necessary for readValue(String content, Class<T> valueType)

### ISSUE THREAD

**Issue by ihr** [FasterXML/jackson-databind#778]
Hi guys,

At present ObjectMapper#readValue(String content, Class<T> valueType) declares IOException. I don't see when an IOException can happen while working with in-memory strings. 

If this is not really need can we discuss removing it? It's really pain in the neck when working with Java 8 lambdas.

**Comment by cowtowncoder** [comment]
While `IOException` should not be thrown, it is not really guaranteed through call stack since access is through regular `Reader`s. But it could be caught, converted to a runtime exception.
But while code could catch `IOException` itself, its sub-classes (`JsonProcessingException`, `JsonParseException`) will still be thrown by underlying parser, databinder.

So benefits of removing `IOException` won't help a lot I think, because `JsonProcessingException` is still exposed similar to other read methods.

One work-around could be to just sub-class `ObjectMapper` as `SafeObjectMapper`, and add wrapper there, to convert `IOException` (and all sub-classes, by extension) into a runtime exception.

**Comment by ihr** [comment]
Hi Tatu,

Yes, I checked the call stack and there several methods that throw `IOException` one of them is in [`JsonFactory#_createParser(Reader r, IOContext ctxt)`](https://github.com/FasterXML/jackson-core/blob/2.5/src/main/java/com/fasterxml/jackson/core/JsonFactory.java#L1207) that can be remove safely.

I like the idea of a `SafeObjectMapper`. Do you need a hand for this issue ? Ii's a small change so I don't know if a pull request is really necessary but if you wish I can do it.

Cheers,
Ivan

**Comment by cowtowncoder** [comment]
Adding `SafeObjectMapper` (or, `UncheckedObjectMapper`) sub-class itself wouldn't be too difficult, but I assume it should convert all checked exceptions while doing that. Reading from `String` is a minority use case, as passing Strings is a very inefficient means to pass JSON content. Because of this, I don't think I would want to add special handling for just this one use case.

Another concern is that while this works for JSON, other sub-classes (like `XmlMapper` for XML) would not be able to use this. For those cases, a delegating variant would make more sense.

Right now the best thing might be for you to locally test out sub-classing, see how that works. Since one of the things we try to do now is to try to limit growth of databind package (there are many things that are useful, but need not be in core packages), possible addition would be good to discuss on dev and/or user lists.

**Comment by ihr** [comment]
On my side I've done what is necessary. I've created this issue only because I think that it will be useful for others and it will be an improvement of the existing code base. You don't have to change anything if you don't agree.

As to reading `String` being minority use case, I don't agree. I'm reading from MongoDB documents in batch of 20 at a time and I need them marshalled once at a time. If you have better proposal than using the aforementioned method, please let me know, I am all ears.

**Comment by cowtowncoder** [comment]
I guess I should have qualified my statement that it is minority use case except for two major cases: unit tests, and using external APIs that insist on exposing JSON as `String`s. :-)
While I consider latter to often be a mistake (since it would be more efficient to use `byte[]`), that's neither here nor there, you can't really change what other systems offer. So your usage is valid, and my comment was misplaced.

I guess I should re-phrase my comment bit differently. Given that other methods that are commonly used for reading JSON can not be changed to eliminate `IOException`, and given that even this method will still need to throw `JsonProcessingException`, I don't think changing signature of that readValue() is worth the hassle. It'd be doable, but I am not sure how valuable it would be.

But as to "safe" `ObjectMapper`, I am open to further ideas on what to do, if anything. The concept seems sound to me if Java 8 lambdas make use of checked exceptions painful (I don't know, haven't used them extensively, but I have heard comments along these lines -- so I assume you are right here). I am not quite sure what to do, however, given some of the challenges. Instead of sub-classing, perhaps delegating model would work well.

I will go ahead and create a new issue for this idea at least, and perhaps ask about this on dev list.

**Comment by ihr** [comment]
Hi,

I agree with you. The most interesting would be the "safe" `ObjectMapper`. 

Let me clarify what the problem with Java 8 lambdas is. Here is the code that we should (ideally) have:

```java
 return findIterable
           .map(result -> mapper.readValue(result.toJson(), target))
           .into(new ArrayList());
```
Instead, with the existing `readValue` method we have:

```java
 return findIterable.map(result -> {
            try {
                return mapper.readValue(result.toJson(), target);
            } catch (IOException e) {
                throw new IllegalStateException(e);
            }
        }).into(new ArrayList<R>());
```

This is very simple but I would also argue quite common use case (although not very efficient way to map large quantities of JSON). At present, the code is polluted unnecessarily which makes it hard to read & maintain. 

Your idea of a delegating model sounds like the right approach to go. I would like to see what do you have in mind exactly. Personally, I would do something along these lins:

```
return findIterable
           .map(result -> mapper.readValue(result.toJson(), target, exceptionHandler))
           .into(new ArrayList());
```

In this way, one has the choice to provide an appropriate `exceptionHandler` per method invocation but why not also have a global handler just in case. This model is similar to what it is in place for Spring framework (in particular the MVC component but also to some extend in the JDBC templating) or to the shutdown hook in Java.

I think other people would be interested in this discussion, so asking for suggestions / comments on the dev list would be helpful. Could you maybe provide a link to the dev list threat in this or the other issue that you are going to create?

Thanks!

**Comment by cowtowncoder** [comment]
Interesting thought on exception handler. That could solve the problem nicely. However, specifically regarding `ObjectMapper` one practical issue is that there are already too many `readValue()` / `writeValue()` methods, so adding one more dimension would be unwieldy. Setting global handler might be better option, although since you can not set/re-set it dynamically, there is some downside.

However: maybe even better option in this would be to instead do that through `ObjectReader` / `ObjectWriter`. Since one can construct new instances cheaply, passing handler would be simple.
That would also only add small number of method calls to support. And in general I am hoping to encourage more use of `ObjectReader`, `ObjectWriter` anyway, since they are fully immutable, thread-safe, reconfigurable (through construction of new instance, not mutability) ways to read/write.

One further benefit here would be that unlike `ObjectMapper`, readers and writers are not currently sub-classes by anything. So delegation model would be less necessary here (it would also be bit more work to support, actually).

So maybe that would be the way to go.

Also, regardless of the approach taken, I think providing a default implementation ("RuntimeException wrapping handler") would make sense.

Great ideas, I think we can get something done here for 2.6 after all.

**Comment by ihr** [comment]
Indeed, `ObjectReader` & `ObjectWriter` seem like good candidates for this work. Furthermore, if this functionality is added to them then people will have one more reason to use them in place of `ObjectMapper` (which is very popular).

Thumbs up for the default implementation for `RuntimeException` wrapping handler (or maybe a bit less generic & precise `IllegalStateException` would be better)

Let me know if I can help with anything for this addition. I'm really looking forward version 2.6 and I'm very excited to have this functionality.

**Comment by cowtowncoder** [comment]
@ihr Sounds good. I did send a message to the dev list, outlining some of challenges.

Agreed also on use of `ObjectReader`/`ObjectWriter`. `ObjectMapper` is quite well entrenched and many users are not even aware of readers/writers, or possible benefits from using. Actually much of my code uses mapper directly already, so there's lots of inertia.

We'll see if this could make it in 2.6 or not. But even if not, one possible option would be a tiny module to provide safe alternative (assuming no support is actually needed by databind), and if it proves useful, could then be merged in 2.7.

When thinking through the problems, I also realized that if I was making bigger changes ("Jackson 3.x"), I'd probably have to refactor mapper, reader and writer, so that there would be abstract base types, and then concrete per-datatype implementations: `JsonMapper`, `JsonReader`, `JsonWriter` (and similarly for other formats). Doing this would allow safe variants to be concrete sub-types at same level as format-variants. Same should be done for current `JsonFactory` as well; make that `CodecFactory` to create `Encoder`/`Decoder` (instead of `JsonGenerator`/`JsonParser`), and then sub-classes for formats (`JsonCodecFactory` / `JsonEncoder` / `JsonDecoder`).
This would help with the open question of how to support format-specific features.

Then again, I don't know if there will be Jackson 3. Due to inertia, Jackson 1.x was (until quite recently) more widely used than Jackson 2.x, and since feature set is sufficient for most, it might be uphill battle to get users to upgrade to another major, backwards-incompatible version.
But I think it makes sense to think of how more optimal structure would look like. Even if it can not be realized currently.
